File: byteback-test/test-java-8/src/main/java/byteback/test/algorithm/DoubleMax.java
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} -o %t.bpl
 */
package byteback.test.algorithm;

import static byteback.specification.Contract.*;
import static byteback.specification.Operators.*;
import static byteback.specification.Quantifiers.*;

import byteback.specification.Bindings;

public class DoubleMax {

	@Behavior
	public static boolean array_is_not_null(double[] a) {
		return neq(a, null);
	}

	@Behavior
	public static boolean max_in_range(double[] a, double n, int start, int end) {
		int i = Bindings.integer();

		return forall(i, implies(lte(start, i) & lt(i, end), gte(n, a[i])));
	}

	@Behavior
	public static boolean array_is_not_empty(double[] a) {
		return gt(a.length, 0);
	}

	@Behavior
	public static boolean result_is_max(double[] a, double t) {
		return max_in_range(a, t, 0, a.length);
	}

	@Require("array_is_not_null")
	@Require("array_is_not_empty")
	@Ensure("result_is_max")
	@Return
	public static double max(double[] a) {
		double t = a[0];

		for (int i = 1; i < a.length; ++i) {
			invariant(lte(0, i) & lte(i, a.length));
			invariant(max_in_range(a, t, 0, i));

			if (a[i] > t) {
				t = a[i];
			}
		}

		return t;
	}

}
/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 2 verified, 0 errors
 */

File: byteback-test/test-scala-2.13.8/src/main/scala/byteback/test/algorithm/DoubleMax.scala
/**
 * RUN: %{byteback} -cp %{jar} -c %{class} --pei --strict -o %t.bpl
 */
package byteback.test.algorithm

import byteback.specification.Contract._
import byteback.specification.Operators._
import byteback.specification.Quantifiers._
import byteback.specification.Bindings

class DoubleMax {

  @Behavior
  def array_is_not_null(a: Array[Double]): Boolean = {
    neq(a, null)
  }

  @Behavior
  def max_in_range(a: Array[Double], n: Double, start: Int, `end`: Int): Boolean = {
    val i = Bindings.integer()

    forall(i, implies(lte(start, i) & lt(i, `end`), gte(n, a(i))))
  }

  @Behavior
  def array_is_not_empty(a: Array[Double]): Boolean = {
    gt(a.length, 0)
  }

  @Behavior
  def result_is_max(a: Array[Double], t: Double): Boolean = {
    max_in_range(a, t, 0, a.length)
  }

  @Require("array_is_not_null")
  @Require("array_is_not_empty")
  @Ensure("result_is_max")
  def max(a: Array[Double]): Double = {
    var t = a(0)
    var i = 1

    while (i < a.length) {
      invariant(lte(0, i) & lte(i, a.length))
      invariant(max_in_range(a, t, 0, i))

      if (a(i) > t) {
        t = a(i)
      }
      i += 1
    }

    t
  }
}
/**
 * RUN: %{verify} %t.bpl | filecheck %s
 * CHECK: Boogie program verifier finished with 2 verified, 0 errors
 */
